src/back/spv/writer.rs

```diff
@@ -885,10 +885,36 @@ impl Writer {
         output: &mut Vec<Instruction>,
     ) -> Instruction {
         match statement {
-            crate::Statement::Return { value: _ } => match function.return_type {
-                Some(_) => unimplemented!(),
-                None => Instruction::new(Op::Return),
-            },
+            crate::Statement::Return { value } => {
+                match function.return_type {
+                    Some(expected_ty) => {
+                        let value = value.expect(
+                            "Return with no value from function supposed to return a value",
+                        );
+
+                        // Parse the expression
+                        let value_expression = &function.expressions[value];
+                        let (value_id, value_ty) =
+                            self.parse_expression(ir_module, function, value_expression, output);
+
+                        // Check that the actual returned type matches the expected one
+                        let expected_ty = &ir_module.types[expected_ty].inner;
+                        assert_eq!(expected_ty, value_ty, "Type of returned expression does not match declared function return type");
```

Comment:
I am not sure if we want add this assert_eq here, but I have to ping you @kvark. This would be a case for issue https://github.com/gfx-rs/naga/issues/59, as the back-end has to expect to get the correct data.